src/compiler/checker.ts

```diff
@@ -16549,6 +16561,40 @@ namespace ts {
             return unknownType;
         }
 
+        // Gets the type of object literal or array literal of destructuring assignment.
+        // { a } from 
+        //     for ( { a } of elems) {
+        //     }
+        // [ a ] from 
+        //     [a] = [ some array ...]
+        function getTypeOfArrayLiteralOrObjectLiteralDestructuringAssignment(expr: Expression): Type {

```

Comment:
Its kind of higher level and gets the type of binding patterns object literal or array literal from binding pattern assignment. ```GetTypeFromBindingPattern``` needs ```BindingPattern```. These are object literals and array literals from assignment so different node kinds